File: core/src/main/java/com/game_machine/game/Gateway.java
package com.game_machine.game;

import java.util.HashMap;

import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;

import com.game_machine.ActorUtil;
import com.game_machine.GameMachine;
import com.game_machine.GatewayMessage;
import com.game_machine.NetMessage;
import com.game_machine.net.server.UdpServer;
import com.game_machine.net.server.UdtServer;

public class Gateway extends UntypedActor {

	LoggingAdapter log = Logging.getLogger(getContext().system(), this);
	private HashMap<String, NetMessage> clients = new HashMap<String, NetMessage>();

	public Gateway() {
	}

	public static void sendToClient(String clientId, byte[] bytes) {
		GatewayMessage gatewayMessage = new GatewayMessage(bytes,clientId);
		ActorUtil.getSelectionByClass(Gateway.class).tell(gatewayMessage, null);	
	}
	
	private String getClientId(NetMessage netMessage) {
		return netMessage.host+":"+netMessage.port;
	}

	public void onReceive(Object message) {
		if (message instanceof NetMessage) {
			NetMessage netMessage = (NetMessage) message;
			GatewayMessage gatewayMessage = new GatewayMessage(netMessage.bytes,getClientId(netMessage));
			clients.put(gatewayMessage.getClientId(), netMessage);
			ActorUtil.getSelectionByClass(GameMachine.getGameHandler()).tell(gatewayMessage, this.getSelf());
			log.debug("Inbound NetMessage message: {}", new String(netMessage.bytes));

		} else if (message instanceof GatewayMessage) {
			GatewayMessage gatewayMessage = (GatewayMessage) message;
			log.debug("Outbound message: {}");
			NetMessage netMessage = clients.get(gatewayMessage.getClientId());

			if (netMessage.protocol == NetMessage.UDP) {
				UdpServer.getUdpServer().send(gatewayMessage.getBytes(), netMessage.host, netMessage.port, netMessage.ctx);
			} else if (netMessage.protocol == NetMessage.UDT) {
				UdtServer.getUdtServer().send(gatewayMessage.getBytes(), netMessage.ctx);
			} else {
				unhandled(message);
			}
		} else {
			unhandled(message);
		}
	}

}

File: ruby/lib/game_machine/gateway.rb
module GameMachine
  class Gateway < UntypedActor

    class << self
      alias_method :apply, :new
      alias_method :create, :new

      def send_to_client(clientId, bytes)
        gatewayMessage = GatewayMessage.new(bytes, clientId)
        ActorUtil.getSelectionByName(self.name).tell(gatewayMessage, nil)
      end

    end

    def initialize
      @clients = {}
    end

    def onReceive(message)
      if message.is_a?(NetMessage)
        handle_net_message(message)
      elsif message.is_a?(GatewayMessage)
        handle_gateway_message(message)
      else 
        unhandled(message)
      end
    end

    private

    def add_client(client_id,net_message)
      @clients[client_id] = net_message
    end

    def get_client(client_id)
      @clients.fetch(client_id)
    end

    def handle_net_message(net_message)
      client_id = get_client_id(net_message)

      gatewayMessage = GatewayMessage.new(net_message.bytes, client_id)
      add_client(client_id,net_message)
      ActorUtil.getSelectionByName(GameMachineLoader.getGameHandler).tell(
        gatewayMessage,
        self.getSelf
      )
      GameMachine.logger.debug("Inbound NetMessage message: #{net_message.bytes}")
    end

    def handle_gateway_message(gateway_message)
      GameMachine.logger.debug("Outbound message: {}")
      net_message = get_client(gateway_message.get_client_id)

      if net_message.protocol == NetMessage::UDP
        UdpServer.get_udp_server.send_to_client(
          gateway_message.get_bytes,
          net_message.host,
          net_message.port,
          net_message.ctx
        )
      elsif net_message.protocol == NetMessage::UDT
        UdtServer.get_udt_server.send_to_client(
          gateway_message.get_bytes,
          net_message.ctx
        )
      else
        unhandled(message)
      end
    end

    def get_client_id(netMessage)
      "#{netMessage.host}:#{netMessage.port}"
    end

  end
end
